jme3-effects/src/main/java/com/jme3/post/filters/ContrastAdjustmentFilter.java

```diff
@@ -0,0 +1,296 @@
+        return lowerLimit;
+    }
+
+    /**
+     * Retrieves the channels input range upper limit.
+     * Default value = 1.0.
+     *
+     * @return the upper limit of the channels input range.
+     */
+    public float getInputRangeUpperLimit() {
+        return upperLimit;
+    }
+
+    /**
+     * Adjusts the scales of different channels.
+     * Default values = 1.0.
+     *
+     * @param redChannelScale   the red channel scale.
+     * @param greenChannelScale the green channel scale.
+     * @param blueChannelScale  the blue channel scale.
+     * @return this filter instance for a chain call.
+     */
+    public ContrastAdjustmentFilter setScales(float redChannelScale, float greenChannelScale, float blueChannelScale) {
+        this.redChannelScale = redChannelScale;
+        this.greenChannelScale = greenChannelScale;
+        this.blueChannelScale = blueChannelScale;
+
+        if (material == null) {
+            return this;
+        }
+
+        //adjust the scales of different channels through the material file
+        material.setFloat("redChannelScale", redChannelScale);
+        material.setFloat("greenChannelScale", greenChannelScale);
+        material.setFloat("blueChannelScale", blueChannelScale);
+        return this;
+    }
+
+    /**
+     * Retrieves the value of the red channel scale that's applied on the final pass.
+     * Default value = 1.0.
+     *
+     * @return the scale of the red channel.
+     */
+    public float getRedChannelScale() {
+        return redChannelScale;
+    }
+
+    /**
+     * Retrieves the value of the green channel scale that's applied on the final pass.
+     * Default value = 1.0.
+     *
+     * @return the scale of the green channel.
+     */
+    public float getGreenChannelScale() {
+        return greenChannelScale;
+    }
+
+    /**
+     * Retrieves the value of the blue channel scale that's applied on the final pass.
+     * Default value = 1.0.
+     *
+     * @return the scale of the blue channel.
+     */
+    public float getBlueChannelScale() {
+        return blueChannelScale;
+    }
+
+    @Override
+    protected void initFilter(AssetManager manager, RenderManager renderManager, ViewPort vp, int w, int h) {
+        //validate app
+        if (manager == null || renderManager == null || vp == null || w == 0 || h == 0) {
+            return;
+        }
+        material = new Material(manager, "Common/MatDefs/Post/ColorContrast.j3md");
+
+        //different channels exp for different transfer functions
+        setExponents(redChannelExponent, greenChannelExponent, blueChannelExponent);
+
+        //input range
+        setInputRange(lowerLimit, upperLimit);
+
+        //final pass scales
+        setScales(redChannelScale, greenChannelScale, blueChannelScale);
+    }
+
+    @Override
+    protected Material getMaterial() {
+        if (material == null) {
+            throw new IllegalStateException("Cannot create a color filter from a null reference !");
```

Comment:
Well I don't know but I want to be safe.

Reply:
Things like calling this factory method from outside this class would make the code fail (ofc almost no one may do this).

Reply:
in this method the '`material`' should always be different from null because it should be initialized in the `initFilter() `method as suggested by the javadoc: 
```java
Initialization of sub classes filtersThis method is called once when the filter is added to the FilterPostProcessorIt should contain Material initializations and extra passes initialization
Overrides: initFilter(...) in Filter
Parameters:
- manager the assetManager
- renderManager the renderManager
- vp the viewPort where this filter is rendered
- w the width of the filter
- h the height of the filter
```

None of the current filters in the com.jme3.post.filters package raise exceptions in this method.

Reply:
The null check should be removed.